README
CRONOLOG FORK version 1.7.1

This is a fork of the original cronolog based on patches in debian &
on the cronolog website. This is only somewhat maintained.

Note: this includes a fix to handle >2GB log files
Note: you can compress on rotate with -G gzip

"cronolog" is a simple program that reads log messages from its input
and writes them to a set of output files, the names of which are
constructed using template and the current date and time.  The
template uses the same format specifiers as the Unix date command
(which are the same as the standard C strftime library function).

"cronolog" is intended to be used in conjunction with a Web server, such
as Apache to split the access log into daily or monthly logs. For
example the Apache configuration directives:

	TransferLog "|/www/sbin/cronolog /www/logs/%Y/%m/%d/access.log"
	ErrorLog    "|/www/sbin/cronolog /www/logs/%Y/%m/%d/errors.log"

would instruct Apache to pipe its access and error log messages into
separate copies of cronolog, which would create new log files each day
in a directory hierarchy structured by date, i.e. on 31 December 1996
messages would be written to

	/www/logs/1996/12/31/access.log
	/www/logs/1996/12/31/errors.log

after midnight the files

	/www/logs/1997/01/01/access.log
	/www/logs/1997/01/01/errors.log

would be used, with the directories 1997, 1997/01 and 1997/01/01 being
created if they did not already exist.

Each character in the template represents a character in the expanded
filename, except for date and time format specifiers, which are
replaced by their expansion.  Format specifiers consist of a `%'
followed by one of the following characters:

 %	a literal % character
 n	a new-line character
 t	a horizontal tab character

Time fields:

 H 	hour (00..23)
 I 	hour (01..12)
 p 	the locale's AM or PM indicator
 M 	minute (00..59)
 S 	second (00..61, which allows for leap seconds)
 X 	the locale's time representation (e.g.: "15:12:47")
 Z 	time zone (e.g. GMT), or nothing if the time zone cannot be determined

Date fields:

 a 	the locale's abbreviated weekday name (e.g.: Sun..Sat)
 A 	the locale's full weekday name (e.g.: Sunday .. Saturday)
 b 	the locale's abbreviated month name (e.g.: Jan .. Dec)
 B 	the locale's full month name, (e.g.: January .. December)
 c 	the locale's date and time (e.g.: "Sun Dec 15 14:12:47 GMT 1996")
 d 	day of month (01 .. 31)
 j 	day of year (001 .. 366)
 m 	month (01 .. 12)
 U 	week of the year with Sunday as first day of week (00..53, where
	week 1 is the week containing the first Sunday of the year)
 W 	week of the year with Monday as first day of week (00..53, where
	week 1 is the week containing the first Monday of the year)
 w 	day of week (0 .. 6, where 0 corresponds to Sunday)
 x 	locale's date representation (e.g. today in Britain: "15/12/96")
 y 	year without the century (00 .. 99)
 Y 	year with the century (1970 .. 2038)

Other specifiers may be available depending on the C library's
implementation of the strftime function.  Before writing a message
cronolog checks the time to see whether the current log file is still
valid and if not it closes the current file, expands the template
using the current date and time to generate a new file name, opens the
new file (creating missing directories on the path of the new log file
as needed unless the program is compiled with -DDONT_CREATE_SUBDIRS)
and calculates the time at which the new file will become invalid.

User specified programs can be run after rotation using the -G param.

The main line cronolog can be found at:

	http://www.ford-mason.co.uk/resources/cronolog/

This fork can be found at:

     	https://github.com/holdenk/cronolog

INSTALLATION
============

Cronolog uses GNU autoconf, so it is configured and built
with the standard sequence:

	./configure
	make

To install type "make install" or simply copy the executable
"src/cronolog" to a suitable directory.

Cronolog has a number of options that can be set either by including
in CFLAGS when configuring, or by editing the file "src/config.h":

	-DFILE_MODE=octal-number	mode used for creating files (default is 0664)
	-DDIR_MODE=octal-number		mode used for creating directories (default is 0775)
	-DDONT_CREATE_SUBDIRS		don't include code to create missing directories
	-DNEED_GETOPT_DEFS		if your platform doesn't declare getopt()

e.g. to set the file mode to 0600:

	CFLAGS="-O2 -DFILE_MODE=0600" ./configure


TESTING
=======

The package includes a simple test harness program (cronotest), which
tests out the lower level functions and prints out debugging
information.  The test program is run with the command line:

	cronotest [-d] "template" count

template is the filename template and count is the number of periods
to evaluate, for example:

	$ cronotest "%Y/%m/%d/access.log" 4
	Determining periodicity of "%Y/%m/%d/access.log"
	%Y -> yearly
	%m -> monthly
	%d -> daily
	Rotation period is per day
	Start time is Sun Dec 15 15:55:43 1996 (850665343)
	Period   1 starts at Sun Dec 15 00:00:00 1996 (850608000):  "1996/12/15/access.log"
	Period   2 starts at Mon Dec 16 00:00:00 1996 (850694400):  "1996/12/16/access.log"
	Period   3 starts at Tue Dec 17 00:00:00 1996 (850780800):  "1996/12/17/access.log"
	Period   4 starts at Wed Dec 18 00:00:00 1996 (850867200):  "1996/12/18/access.log"

Specifying the flag -d instructs the test program to create missing
subdirectories, for example, having created a directory 1996:

	$ cronotest -d  "%Y/%m/%d/access.log" 4
	Determining periodicity of "%Y/%m/%d/access.log"
	%Y -> yearly
	%m -> monthly
	%d -> daily
	Rotation period is per day
	Start time is Sun Dec 15 16:00:21 1996 (850665621)
	Period   1 starts at Sun Dec 15 00:00:00 1996 (850608000):  "1996/12/15/access.log"
	Creating missing components of 1996/12/15/access.log
	Testing directory "1996"
	Testing directory "1996/12"
	Directory "1996/12" does not exist -- creating
	Testing directory "1996/12/15"
	Directory "1996/12/15" does not exist -- creating
	Period   2 starts at Mon Dec 16 00:00:00 1996 (850694400):  "1996/12/16/access.log"
	Creating missing components of 1996/12/16/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/16
	Directory "1996/12/16" does not exist -- creating
	Period   3 starts at Tue Dec 17 00:00:00 1996 (850780800):  "1996/12/17/access.log"
	Creating missing components of 1996/12/17/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/17
	Directory "1996/12/17" does not exist -- creating
	Period   4 starts at Wed Dec 18 00:00:00 1996 (850867200):  "1996/12/18/access.log"
	Creating missing components of 1996/12/18/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/18
	Directory "1996/12/18" does not exist -- creating


Extra debugging code was added to cronolog at version 1.5b8 to
facilitate interactive testing.  To activate this cronolog must be
must be invoked with -x (use -x- to send debug messages to stderr).
In this mode each time it reads a message cronolog will print a
message giving the current time, the start of the next period and the
number of seconds until the start of the next period.  This option is

debian/changelog
cronolog (1.7.2) unstable; urgency=low

  * Fix buffer overflow
    fix 2GB file issue
    add rolling gzip support

 -- Holden Karau <holden@pigscanfly.ca>  Fri, 25 Apr 2014 21:23:20 -0800

cronolog (1.7-1) unstable; urgency=low

  * Merge in all the patches
    fix 2GB file issue
    add rolling gzip support

 -- Holden Karau <holden@pigscanfly.ca>  Wed, 23 Apr 2014 16:57:59 -0800

cronolog (1.6.2+rpk-1) unstable; urgency=low

  * RePacKaging of orig.tar.gz to change source tarball in the archive.
    (was a tarball-in-tarball source archive, for dbs use)

 -- Maxime Chatelle <mmyc@gmx.com>  Tue, 14 Jun 2011 23:47:39 +0200

cronolog (1.6.2-6) unstable; urgency=low

  * New Maintainer: Maxime Chatelle <mmyc@gmx.com> (Closes: #465934).
  * Upgraded to Policy 3.9.2.
  * Upgraded to "3.0 (quilt)" package format and full rewrite of 
    debian/rules, removing dbs build dependency (Closes: #576045).
  * cronolog manpage: typo error fixed (050_cronolog_manpage_typo.diff)
    (Closes: #217554).
  * cronosplit manpage: add some precisions about supported formats
    (060_cronosplit_manpage_correction.diff)(Closes: #217566).
  * removed "030_define_xopen_source.diff" patch and use CFLAGS in
    debian/rules instead (see 1.6.2-5.1, was #323425).
  * added -D_FILE_OFFSET_BITS=64 in CFLAGS for large file support
    (Closes: #304626).
  * removed the, now very old, wrapper in /usr/sbin (see 1.6.2-3).
  * errors fixed in manpages (070_manpages_fixes.diff)
  * --utime option added to cronosplit (Closes: #370297).

 -- Maxime Chatelle <mmyc@gmx.com>  Mon, 4 Jun 2011 14:49:57 +0200

cronolog (1.6.2-5.3) unstable; urgency=low

  * Porter NMU.
  * Update config.guess/config.sub at build-time (Closes: #331428).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 10 Dec 2006 23:45:56 +0100

cronolog (1.6.2-5.2) unstable; urgency=low

  * Non-maintainer upload.
  * cronosplit bugfixes:
    * Print content of invalid lines.
    * Also split error logfiles. (Closes: #241492)
    * Safe open for input files/pipes.
  * cronosplit improvements:
    * use strict;
    * Use strftime for template conversion.
    * Add processing of bzip2 compressed logfiles.
  * Incorporate the above changes + 020_cronosplit_HMS.diff into new
    040_cronosplit.diff.

 -- Christoph Biedl <cbiedl@gmx.de>  Sun,  4 Jun 2006 13:20:49 +0200

cronolog (1.6.2-5.1) unstable; urgency=low

  * Non-maintainer upload.
  * Define _XOPEN_SOURCE to fix an implicitly converted pointer.
    Closes: #323425

 -- dann frazier <dannf@debian.org>  Tue, 27 Sep 2005 00:36:07 -0600

cronolog (1.6.2-5) unstable; urgency=low

  * Upgraded to Policy 3.6.1.
  * Corrected patch 020_cronosplit_HMS.diff to also patch cronosplit.in
    (we don't regenerate anything, so it isn't crucial, but it's good to
    keep things in sync).
  * Added upstream ChangeLog, NEWS, README, and TODO files to the doc
    directory. (Closes: #241487)

 -- Joel Baker <fenton@debian.org>  Mon,  5 Apr 2004 20:54:01 -0600

cronolog (1.6.2-4) unstable; urgency=low

  * Upgraded to Policy 3.6.0.
  * Fixed a problem in cronosplit which caused %M and %S templates to go
    haywire. (Closes: #203372)

 -- Joel Baker <fenton@debian.org>  Tue, 29 Jul 2003 22:44:10 -0600

cronolog (1.6.2-3) unstable; urgency=low

  * Changed documentation to point to /usr/bin, and added a wrapper
    script in /usr/sbin that throws a message to STDERR warning about
    the deprecated location. The wrapper can be dropped after the next
    major release. (Closes: #193822)
  * Patch cronolog.info to be in section 'Administrative commands'; next
    upstream release will put it there anyway. (Closes: #193350)


src/cronolog.c
/* ====================================================================
 * Copyright (c) 1995-1999 The Apache Group.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All advertising materials mentioning features or use of this
 *    software must display the following acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * 4. The names "Apache Server" and "Apache Group" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For written permission, please contact
 *    apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache"
 *    nor may "Apache" appear in their names without prior written
 *    permission of the Apache Group.
 *
 * 6. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * THIS SOFTWARE IS PROVIDED BY THE APACHE GROUP ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE APACHE GROUP OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Group and was originally based
 * on public domain software written at the National Center for
 * Supercomputing Applications, University of Illinois, Urbana-Champaign.
 * For more information on the Apache Group and the Apache HTTP server
 * project, please see <http://www.apache.org/>.
 *
 */
/*
 * cronolog -- simple Apache log rotation program
 *
 * Copyright (c) 1996-1999 by Ford & Mason Ltd
 *
 * This software was submitted by Ford & Mason Ltd to the Apache
 * Software Foundation in December 1999.  Future revisions and
 * derivatives of this source code must acknowledge Ford & Mason Ltd
 * as the original contributor of this module.  All other licensing
 * and usage conditions are those of the Apache Software Foundation.
 *
 * Originally written by Andrew Ford <A.Ford@ford-mason.co.uk>
 *
 * cronolog is loosly based on the rotatelogs program, which is part of the
 * Apache package written by Ben Laurie <ben@algroup.co.uk>
 *
 * The argument to this program is the log file name template as an
 * strftime format string.  For example to generate new error and
 * access logs each day stored in subdirectories by year, month and day add
 * the following lines to the httpd.conf:
 *
 *    TransferLog "|/www/etc/cronolog /www/logs/%Y/%m/%d/access.log"
 *    ErrorLog    "|/www/etc/cronolog /www/logs/%Y/%m/%d/error.log"
 *
 * The option "-x file" specifies that debugging messages should be
 * written to "file" (e.g. /dev/console) or to stderr if "file" is "-".
 */

#ifndef _WIN32
#define _GNU_SOURCE 1
#define OPEN_EXCLUSIVE O_WRONLY|O_CREAT|O_EXCL|O_APPEND|O_LARGEFILE
#define OPEN_SHARED O_WRONLY|O_CREAT|O_APPEND|O_LARGEFILE
#else
#define OPEN_EXCLUSIVE O_WRONLY|O_CREAT|O_EXCL|O_APPEND
#define OPEN_SHARED O_WRONLY|O_CREAT|O_APPEND
#endif

#include "config.h"

#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>
#include "cronoutils.h"
#include "getopt.h"


/* Forward function declaration */

int    new_log_file(const char *, const char *, mode_t, const char *,
             PERIODICITY, int, int, char *, size_t, time_t, time_t *);

void    cleanup(int );
void    handle_file();
void    fork_to_handle_file();

int     openwrapper( const char *filename );

#ifndef _WIN32
void    setsig_handler( int signum,  void (*action)(int, siginfo_t *, void *));
void    set_signal_handlers();
#endif

/* Definition of version and usage messages */

#ifndef _WIN32
#define VERSION_MSG       PACKAGE " version " VERSION "\n"
#else
#define VERSION_MSG      "cronolog version 0.1\n"
#endif

#ifndef _WIN32
#define SETUGID_USAGE    "   -u USER,   --set-uid=USER  change to USER before doing anything (name or UID)\n" \
            "   -g GROUP,  --set-gid=GROUP change to GROUP before doing anything (name or GID)\n"
#else
#define SETUGID_USAGE    ""
#endif

#define USAGE_MSG     "usage: %s [OPTIONS] logfile-spec\n" \
            "\n" \
            "   -H NAME,   --hardlink=NAME maintain a hard link from NAME to current log\n" \
            "   -S NAME,   --symlink=NAME  maintain a symbolic link from NAME to current log\n" \
            "   -P NAME,   --prev-symlink=NAME  maintain a symbolic link from NAME to previous log\n" \
            "   -l NAME,   --link=NAME     same as -S/--symlink\n" \
            "   -h,        --help          print this help, then exit\n" \
            "   -p PERIOD, --period=PERIOD set the rotation period explicitly\n" \
            "   -d DELAY,  --delay=DELAY   set the rotation period delay\n" \
            "   -o,        --once-only     create single output log from template (not rotated)\n" \
            "   -x FILE,   --debug=FILE    write debug messages to FILE\n" \
            "                              ( or to standard error if FILE is \"-\")\n" \
            "   -r,        --helper=SCRIPT post rotation helper script to fork exec on old files\n" \
            "                              ( will be called like \"SCRIPT <oldlog>\" )\n" \
            "                              ( not tested on windows )\n" \
            "   -G,        --helper-arg=ARG argument passed to rotation helper script\n" \
            SETUGID_USAGE \
            "   -a,        --american         American date formats\n" \
            "   -e,        --european         European date formats (default)\n" \
            "   -n,        --no-alarm      wait to rotate logs and create new ones until there is traffic\n" \
            "   -s,    --start-time=TIME   starting time\n" \
            "   -z TZ, --time-zone=TZ      use TZ for timezone\n" \
            "   -V,      --version         print version number, then exit\n"


/* Definition of the short and long program options */

char          *short_options = "ad:enop:s:z:H:P:S:l:hVx:r:G:u:g:M:D:";

#ifndef _WIN32
struct option long_options[] =
{
    { "american",    no_argument,        NULL, 'a' },
    { "european",    no_argument,        NULL, 'e' },
    { "start-time",     required_argument,    NULL, 's' },
    { "time-zone",      required_argument,    NULL, 'z' },
    { "hardlink",      required_argument,     NULL, 'H' },
    { "symlink",       required_argument,     NULL, 'S' },
    { "prev-symlink",      required_argument,     NULL, 'P' },
    { "link",          required_argument,     NULL, 'l' },
    { "period",        required_argument,    NULL, 'p' },
    { "delay",        required_argument,    NULL, 'd' },
    { "helper",        required_argument,    NULL, 'r' },
    { "helper-arg",    required_argument,    NULL, 'G' },
    { "set-uid",    required_argument,  NULL, 'u' },
    { "set-gid",    required_argument,  NULL, 'g' },
    { "once-only",     no_argument,           NULL, 'o' },
    { "no-alarm",     no_argument,           NULL, 'n' },
    { "lock",         no_argument,           NULL, 'L' },
    { "file-mode",     required_argument,           NULL, 'M' },
    { "dir-mode",     required_argument,           NULL, 'D' },
    { "help",          no_argument,           NULL, 'h' },
    { "version",       no_argument,           NULL, 'V' }
};
#endif

static    char    handler[MAX_PATH+1];
static    char    handler_arg[MAX_PATH+1];
static    char    filename[MAX_PATH+1];
//  .happypandas is 13 chars and null
static    char    lockfilename[MAX_PATH+14];
static    int   use_handler =0;
static    int   use_handler_arg =0;
static    int   i_am_handler =0;
int     acquire_lock = 0;

//
// arguably we should use mode_t, but obscuring the type
// here can only hurt, since we have to specify a scanf format 
// string later on. 
//
static    unsigned int   file_mode = FILE_MODE ;
static    unsigned int   dir_mode = DIR_MODE ;

/* Main function.
 */
int
main(int argc, char **argv)
{
    PERIODICITY    periodicity = UNKNOWN;
    PERIODICITY    period_delay_units = UNKNOWN;
    int        period_multiple = 1;
    int        period_delay  = 0;
    int        use_american_date_formats = 0;
    char     read_buf[BUFSIZE];
    char     tzbuf[BUFSIZE];
    char    *start_time = NULL;
    char    *template;
    char    *linkname = NULL;
    char    *prevlinkname = NULL;
#ifndef _WIN32
    uid_t    new_uid = 0;
    gid_t    new_gid = 0;
    int        change_uid = 0;
    int        change_gid = 0;
#endif
    mode_t    linktype = 0;
    int     n_bytes_read;
    int        ch;
    time_t    time_now;
    time_t    time_offset = 0;
    time_t    next_period = 0;
    int     log_fd = -1;
    int     no_alarm = 0;

    memset( handler, '\0', MAX_PATH+1 );
    memset( handler_arg, '\0', MAX_PATH+1 );
    memset( filename, '\0', MAX_PATH+1 );
    // 13 is length of .happypandas , 14 with \0 :)
    memset( lockfilename, '\0', MAX_PATH+14 );

#ifndef _WIN32
    while ((ch = getopt_long(argc, argv, short_options, long_options, NULL)) != EOF)
#else
    while ((ch = getopt(argc, argv, short_options)) != EOF)
#endif        
    {
    switch (ch)
    {
    case 'a':
        use_american_date_formats = 1;
        break;
        
    case 'e':
        use_american_date_formats = 0;
        break;
        
    case 's':
        start_time = optarg;
        break;

    case 'z':
        sprintf(tzbuf, "TZ=%s", optarg);
        putenv(tzbuf);
        break;

    case 'H':
        linkname = optarg;
        linktype = S_IFREG;
        break;

    case 'l':
    case 'S':
        linkname = optarg;
#ifndef _WIN32
        linktype = S_IFLNK;
#endif        
        break;
    case 'L':
        acquire_lock = 1;
        break;
    case 'P':
        if (linkname == NULL)
        {
        fprintf(stderr, "A current log symlink is needed to mantain a symlink to the previous log\n");
        exit(1);
        }
        prevlinkname = optarg;
        break;
        

    case 'd':
        period_delay_units = parse_timespec(optarg, &period_delay);
        break;

    case 'p':
        periodicity = parse_timespec(optarg, &period_multiple);
        if (   (periodicity == INVALID_PERIOD)
        || (periodicity == PER_SECOND) && (60 % period_multiple)
        || (periodicity == PER_MINUTE) && (60 % period_multiple)
        || (periodicity == HOURLY)     && (24 % period_multiple)
        || (periodicity == DAILY)      && (period_multiple > 365)
        || (periodicity == WEEKLY)     && (period_multiple > 52)
        || (periodicity == MONTHLY)    && (12 % period_multiple)) {
        fprintf(stderr, "%s: invalid explicit period specification (%s)\n", argv[0], start_time);
        exit(1);
        }        
        break;
        
#ifndef _WIN32
    case 'u':
        new_uid = parse_uid(optarg, argv[0]);
        change_uid = 1;
        break;
    case 'g':
        new_gid = parse_gid(optarg, argv[0]);
        change_gid = 1;
        break;
#endif
    case 'M':
        sscanf( optarg, "%o", & file_mode );
        break;
    case 'D':
        sscanf( optarg, "%o", & dir_mode );
        break;
    case 'o':
        periodicity = ONCE_ONLY;
        break;
        
    case 'x':
        if (strcmp(optarg, "-") == 0)
        {
        debug_file = stderr;
        }
        else
        {
        debug_file = fopen(optarg, "a+");
        }
        break;
    case 'r':
            strncat(handler, optarg, MAX_PATH - strlen(handler) );
            use_handler=1;
            break;
    case 'G':
            strncat(handler_arg, optarg, MAX_PATH - strlen(handler) );
            use_handler_arg=1;
            break;
    case 'n':
            no_alarm=1;
            break;

    case 'V':
        fprintf(stderr, VERSION_MSG);
        exit(0);
        
    case 'h':
    case '?':
        fprintf(stderr, USAGE_MSG, argv[0]);
        exit(1);
    }
    }

    if ((argc - optind) != 1)
    {
    fprintf(stderr, USAGE_MSG, argv[0]);
    exit(1);
    }

#ifndef _WIN32
    if (change_gid && 
#ifndef CRONOLOG_HAS_SETREUID
            setgid(new_gid) 
#else 
            setregid(new_gid,new_gid) 
#endif  /* CRONOLOG_HAS_SETEUID */
            == -1) {
    fprintf(stderr, "setgid: unable to change to gid: %d\n", new_gid);
           exit(1);
    }
    if (change_uid && 
#ifndef CRONOLOG_HAS_SETREUID
            setuid(new_uid) 
#else 
            setreuid(new_uid,new_uid) 
#endif  /* CRONOLOG_HAS_SETEUID */
            == -1) {
    fprintf(stderr, "setuid: unable to change to uid: %d\n", new_uid);
           exit(1);
    }
#endif

    DEBUG((VERSION_MSG "\n"));

    if (start_time)
    {
    time_now = parse_time(start_time, use_american_date_formats);
    if (time_now == -1)
    {
        fprintf(stderr, "%s: invalid start time (%s)\n", argv[0], start_time);
        exit(1);
    }
    time_offset = time_now - time(NULL);
    DEBUG(("Using offset of %d seconds from real time\n", time_offset));
    }

    /* The template should be the only argument.
     * Unless the -o option was specified, determine the periodicity.
     */
    
    template = argv[optind];
    if (periodicity == UNKNOWN)
    {
    periodicity = determine_periodicity(template);
    }


    DEBUG(("periodicity = %d %s\n", period_multiple, periods[periodicity]));

    if (period_delay) {
    if (   (period_delay_units > periodicity)
        || (   period_delay_units == periodicity
        && abs(period_delay)  >= period_multiple)) {
        fprintf(stderr, "%s: period delay cannot be larger than the rollover period\n", argv[0]);
        exit(1);
    }        
    period_delay *= period_seconds[period_delay_units];
    }

    DEBUG(("Rotation period is per %d %s\n", period_multiple, periods[periodicity]));


#ifndef _WIN32
    set_signal_handlers();

    if( ! no_alarm )
    {
        //ensure that the first read() does not hang, and that the log file is promptly created
        //subsequent alarm() calls will set the alarm for the rotation period.
        alarm( 1 ); 
    }
#endif


    /* Loop, waiting for data on standard input */

    for (;;)
    {
    /* Read a buffer's worth of log file data, exiting on errors
     * or end of file.
     */
    n_bytes_read = read(0, read_buf, sizeof read_buf);
    if (n_bytes_read == 0)
    {
        cleanup(3);
    }
    if (errno == EINTR)
    {
        /* 
             * fall through, it may have been alarm, in which case it will be time to rotate.
             * */
    }
    else if (n_bytes_read < 0)
    {
        cleanup(4);
    }

    time_now = time(NULL) + time_offset;
    
    /* If the current period has finished and there is a log file
     * open, close the log file
     */
    if ((time_now >= next_period) && (log_fd >= 0))
    {
        close(log_fd);
        log_fd = -1;
            fork_to_handle_file();
    }
    
    /* If there is no log file open then open a new one.
     */
    if (log_fd < 0)
    {
        log_fd = new_log_file(template, linkname, linktype, prevlinkname,
                  periodicity, period_multiple, period_delay,
                  filename, sizeof (filename), time_now, &next_period);
#ifndef _WIN32
        if( ! no_alarm )
        {
            alarm( next_period - time_now );
        }
#endif
    }

    DEBUG(("%s (%d): wrote message; next period starts at %s (%d) in %d secs\n",
           timestamp(time_now), time_now, 
           timestamp(next_period), next_period,
           next_period - time_now));

    /* Write out the log data to the current log file.
     */
    if (n_bytes_read && write(log_fd, read_buf, n_bytes_read) != n_bytes_read)
    {
        perror(filename);
        cleanup(5);
    }
    }

    /* NOTREACHED */
    return 1;
}

/* Open a new log file: determine the start of the current
 * period, generate the log file name from the template,
 * determine the end of the period and open the new log file.
 *
 * Returns the file descriptor of the new log file and also sets the
 * name of the file and the start time of the next period via pointers
 * supplied.
 */
int
new_log_file(const char *template, const char *linkname, mode_t linktype, const char *prevlinkname,
         PERIODICITY periodicity, int period_multiple, int period_delay,
         char *pfilename, size_t pfilename_len,
         time_t time_now, time_t *pnext_period)
{
    time_t     start_of_period;
    struct tm     *tm;
    int     log_fd;


    start_of_period = start_of_this_period(time_now, periodicity, period_multiple);
    tm = localtime(&start_of_period);
    strftime(pfilename, BUFSIZE, template, tm);
    *pnext_period = start_of_next_period(start_of_period, periodicity, period_multiple) + period_delay;
    
    DEBUG(("%s (%d): using log file \"%s\" from %s (%d) until %s (%d) (for %d secs)\n",
    timestamp(time_now), time_now, pfilename, 
       timestamp(start_of_period), start_of_period,
    timestamp(*pnext_period), *pnext_period,
    *pnext_period - time_now));
    
    log_fd = openwrapper(pfilename);
    
#ifndef DONT_CREATE_SUBDIRS
    if ((log_fd < 0) && (errno == ENOENT))
    {
        create_subdirs(pfilename, dir_mode);
        log_fd = openwrapper(pfilename);
    }
#endif        

    if (log_fd < 0)
    {
        perror(pfilename);
        exit(2);
    }

    if (linkname)
    {
        create_link(pfilename, linkname, linktype, prevlinkname);
    }
    return log_fd;
}

/* 
 * fork, then exec an external handler to deal with rotated file.
 */
void
fork_to_handle_file()
{
    int fk ;
    static int childpid=0;

    if( ! use_handler || !i_am_handler || handler[0] =='\0' || filename[0] == '\0' )
    {
        return;
    }
    fk=fork();
    if( fk < 0 )
    {
        perror("couldnt fork");
        exit(2);
    }else if( fk > 0 )
    {
        if( childpid )
        {
            /* 
             * collect zombies. run twice, in case one or more children took longer than
             * the rotation period for a while, this will eventually clean them up.
             * Of course, if handler children take longer than rotation period to handle
             * things, you will eventually have a big problem.
             * 
             * */ 
            (void) waitpid( 0, NULL, WNOHANG | WUNTRACED );
            (void) waitpid( 0, NULL, WNOHANG | WUNTRACED );
        }
        childpid=fk;
        return; /* parent */
    }
    /* child */
    /* dont muck with stdin or out of parent, but allow stderr to be commingled */
    close(0);
    close(1);
    handle_file();
}

/* 
 * exec an external handler to deal with rotated file.
 */
void 
handle_file()
{
    char **exec_argv ;
    char **exec_envp ;
    if( ! use_handler || !i_am_handler || handler[0] =='\0' || filename[0] == '\0' )
    {
        return;
    }
    if ( acquire_lock ) {
      strncpy(lockfilename, filename, MAX_PATH);
      strcat(lockfilename, ".happypandas");
      int ret = open(lockfilename, OPEN_EXCLUSIVE, 'w');
      // file already exists
      if (ret == -1) {
	exit(0);
      }
    }
    exec_envp = malloc( sizeof( char *)*1);
    exec_envp[0] = NULL;

    exec_argv = malloc( sizeof( char *)*6);
    int i=0;
    //exec_argv[i++] = strdup( "/bin/bash" );
    //exec_argv[i++] = strdup( "--login" );
    exec_argv[i++] = strdup( handler );
    if ( use_handler_arg == 1 )
    {
        exec_argv[i++] = strdup( handler_arg );
    }
    exec_argv[i++] = strdup( filename );
    exec_argv[i++] = NULL;

    execve( exec_argv[0], exec_argv, exec_envp );
    perror("cant execvp");
    exit(2);
}



#ifndef _WIN32
/* 
 * wrapper to be called as signal handler.
 */
void 
handle_file_on_sig( int sig, siginfo_t *si, void *v )
{
    handle_file();
    /* not reached */
    exit( 3 );
};

/* 
 * wrapper to be called for alarm signal
 */
void 
alarm_signal_handler( int sig, siginfo_t *si, void *v )
{
        ;
        /* 
         * do nothing; the key thing is that the alarm will cause the read()
         * to fail with errno=EINTR. this empty handler is required, because the
         * default handler will exit(1)
         *
         */
};

void
set_signal_handlers()
{
    /* 
     * all signals which usually kill a process that can be caught are
     * set to handle_file when received. This will make apache shutdowns more 
     * graceful even if use_handler is false.
     */
    setsig_handler( SIGHUP, handle_file_on_sig );
    setsig_handler( SIGINT, handle_file_on_sig );
    setsig_handler( SIGQUIT, handle_file_on_sig );
    setsig_handler( SIGILL, handle_file_on_sig );
    setsig_handler( SIGABRT, handle_file_on_sig );
    setsig_handler( SIGBUS, handle_file_on_sig );
    setsig_handler( SIGFPE, handle_file_on_sig );
    setsig_handler( SIGPIPE, handle_file_on_sig );
    setsig_handler( SIGTERM, handle_file_on_sig );
    setsig_handler( SIGUSR1, handle_file_on_sig );

    /* sigalrm is used to break out of read() when it is time to rotate the log. */
    setsig_handler( SIGALRM, alarm_signal_handler );
}

void
setsig_handler( int signum,  void (*action)(int, siginfo_t *, void *))
{
    struct sigaction siga ;
    memset( &siga, '\0', sizeof( struct sigaction ));
    siga.sa_sigaction= action ;
    siga.sa_flags = SA_SIGINFO ;
    if( -1== sigaction( signum, &siga, NULL ))
    {
        perror( "cant set sigaction" );
    }
}
#endif


/* 
 * cleanup
 */
void
cleanup( int exit_status )
{
    handle_file();
    exit(exit_status);
}

/* 
 * only the first cronolog process to open a particular file is responsible
 * for starting the cleanup process later. This wrapper sets i_am_handler
 * according to that logic.
 * */
int 
openwrapper( const char *ofilename )
{
    int ret;
    if( use_handler !=1 )
    {
        return open(ofilename, OPEN_SHARED, file_mode );
    }
    ret = open(ofilename, OPEN_EXCLUSIVE, file_mode );
    if( ret < 0 )
    {
        ret = open(ofilename, OPEN_SHARED, file_mode );
        i_am_handler= 0;
    }
    else
    {
        i_am_handler=1;
    }
    return ret;
}

